File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // hash table
        unordered_map <string, vector<string>> hmap;

        vector<vector<string>> result;

        // create a copy of input vector
        vector<string> strscpy = strs;

        // sorting each element in copy vector
        for(int i=0; i<strscpy.size(); i++) {
            sort(strscpy[i].begin(), strscpy[i].end());
            hmap[strscpy[i]].push_back(strs[i]);
        }

        // using hash table to get back the final output
        for (auto k: hmap) result.push_back(k.second);

        return result;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        dir = {}
        
        def dna(s: str) -> str:
            key = ""
            dir = [0] * 26
            for ch in s:
                dir[ord(ch)-ord("a")] += 1

            for idx, count in enumerate(dir):
                if count: 
                    key += chr(idx)+str(count)

            return key

        for s in strs:
            k = dna(s)
            if dir.get(k):
                dir[k].append(s)
            else:
                dir[k] = [s]


        return list(dir.values())